File: problems/CSES/Graph_Algorithms/building_roads/building_roads.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n + 1)]

    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    visited = [False] * (n + 1)
    representatives = []

    def dfs(start):
        stack = [start]
        visited[start] = True

        while stack:
            node = stack.pop()

            for nei in graph[node]:
                if not visited[nei]:
                    visited[nei] = True
                    stack.append(nei)

    for city in range(1, n + 1):
        if not visited[city]:
            representatives.append(city)
            dfs(city)

    print(len(representatives) - 1)

    for i in range(1, len(representatives)):
        print(representatives[i - 1], representatives[i])

main()

File: problems/CSES/Graph_Algorithms/building_roads/building_roads.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    vector<bool> visited(n + 1, false);
    vector<int> representatives;
    vector<int> stack;

    for (int city = 1; city <= n; ++city) {
        if (!visited[city]) {
            representatives.push_back(city);
            
            stack.push_back(city);
            visited[city] = true;

            while (!stack.empty()) {
                int node = stack.back();
                stack.pop_back();

                for (int nei : graph[node]) {
                    if (!visited[nei]) {
                        visited[nei] = true;
                        stack.push_back(nei);
                    }
                }
            }
        }
    }

    cout << representatives.size() - 1 << "\n";

    for (size_t i = 1; i < representatives.size(); ++i) {
        cout << representatives[i - 1] << " " << representatives[i] << "\n";
    }

    return 0;
}
